node/src/reactor/joiner.rs

```diff
@@ -978,6 +1066,210 @@ impl Reactor {
         self.rest_server.finalize().await;
         Ok(config)
     }
+
+    // If we have a latest block header, we joined via the joiner reactor.
+    //
+    // If not, run genesis or upgrade and construct a switch block, and use that for the latest
+    // block header.
+    async fn commit_upgrade_or_genesis_switch_block(
+        maybe_latest_block_header: Option<BlockHeader>,
+        chainspec: Arc<Chainspec>,
+        storage: Arc<Mutex<Storage>>,
+        contract_runtime: &ContractRuntime,
+        effect_builder: EffectBuilder<JoinerEvent>,
+    ) -> Result<BlockHeader, Error> {
+        let block_header = match maybe_latest_block_header {
+            Some(latest_block_header)
+                if latest_block_header.protocol_version() == chainspec.protocol_config.version =>
+            {
+                latest_block_header
+            }
+            Some(latest_block_header)
+                if chainspec
+                    .protocol_config
+                    .is_last_block_before_activation(&latest_block_header) =>
+            {
+                Self::commit_upgrade_switch_block(
+                    latest_block_header,
+                    chainspec,
+                    storage,
+                    contract_runtime,
+                    effect_builder,
+                )
+                .await?
+            }
+            Some(latest_block_header) => {
+                return Err(Error::UnexpectedLatestBlockHeader {
+                    latest_block_header: Box::new(latest_block_header),
+                });
+            }
+            None => match chainspec.protocol_config.activation_point {
+                ActivationPoint::EraId(upgrade_era_id) => {
+                    return Err(Error::NoSuchSwitchBlockHeaderForUpgradeEra { upgrade_era_id });
+                }
+                ActivationPoint::Genesis(genesis_timestamp) => {
+                    // Do not run genesis on a node which is not protocol version 1.0.0
+                    Self::commit_genesis_switch_block(
+                        chainspec,
+                        storage,
+                        contract_runtime,
+                        effect_builder,
+                        genesis_timestamp,
+                    )
+                    .await?
+                }
+            },
+        };
+        Ok(block_header)
```

Comment:
No need for the local `block_header` variable.

Reply:
Done